File: app/src/main/java/com/hundaol/ethiopic/MainActivity.java
package com.hundaol.ethiopic;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.view.Display;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.hundaol.ethiocal.R;
import com.hundaol.ethiopic.cal.GregorianCal;
import com.hundaol.ethiopic.domain.DateModel;
import com.jakewharton.rxbinding2.view.RxView;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.disposables.CompositeDisposable;

public class MainActivity extends AppCompatActivity {

    @Inject
    FirebaseAnalytics firebaseAnalytics;

    @Inject
    Display display;

    @Inject
    DateModel dateModel;

    @BindView(R.id.view_pager)
    ViewPager viewPager;

    @BindView(R.id.today)
    FloatingActionButton today;

    MainPagerAdapter pagerAdapter;

    private final CompositeDisposable disposables = new CompositeDisposable();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        App.getAppComponent().inject(this);

        setContentView(R.layout.activity_main);

        ButterKnife.bind(this);

        if (!checkIfAlreadyHavePermission()) {
            requestForSpecificPermission();
        }

        pagerAdapter = new MainPagerAdapter(this, display, dateModel);
        viewPager.setAdapter(pagerAdapter);

        viewPager.setCurrentItem(1);
    }

    @Override
    protected void onResume() {
        super.onResume();
        disposables.add(RxView.clicks(today).subscribe(v -> dateModel.setJdv(GregorianCal.INSTANCE.today())));
    }

    @Override
    protected void onPause() {
        super.onPause();
        disposables.dispose();
    }

    private void requestForSpecificPermission() {
        ActivityCompat.requestPermissions(this,
                new String[]{Manifest.permission.READ_CALENDAR},
                1);
    }

    private boolean checkIfAlreadyHavePermission() {
        int result = ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CALENDAR);
        if (result == PackageManager.PERMISSION_GRANTED) {
            return true;
        } else {
            return false;
        }
    }

}

File: app/src/main/java/com/hundaol/ethiopic/MainActivity.kt
package com.hundaol.ethiopic

import android.Manifest
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.support.v4.view.PagerAdapter
import android.support.v4.view.ViewPager
import android.support.v7.app.AppCompatActivity
import android.view.Display

import com.google.firebase.analytics.FirebaseAnalytics
import com.hundaol.ethiocal.R

import javax.inject.Inject

import butterknife.BindView
import butterknife.ButterKnife
import com.hundaol.ethiopic.cal.GregorianCal
import com.jakewharton.rxbinding2.view.RxView
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber

class MainActivity : AppCompatActivity() {

    @Inject
    lateinit var firebaseAnalytics: FirebaseAnalytics

    @Inject
    lateinit var display: Display

    @BindView(R.id.view_pager)
    lateinit var viewPager: ViewPager

    @BindView(R.id.today)
    lateinit var todayButton: FloatingActionButton

    lateinit var pagerAdapter: PagerAdapter

    private val disposables = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        App.appComponent.inject(this)

        setContentView(R.layout.activity_main)

        ButterKnife.bind(this)

        if (!hasPermissionReadCalendar()) {
            requestPermissionReadCalendar()
        }

        pagerAdapter = MainPagerAdapter(this, display)
        viewPager.adapter = pagerAdapter
        viewPager.currentItem = 1
    }

    override fun onResume() {
        super.onResume()
        disposables.add(RxView.clicks(todayButton).subscribe(
                { v ->
                    App.setJdv(GregorianCal.INSTANCE.today().toFloat(), 1000L);
                },
                { error ->
                    Timber.w(error, "error observed on todayButton clicks subscription")
                }
        ))
    }

    override fun onPause() {
        super.onPause()
        disposables.dispose()
    }

    fun requestPermissionReadCalendar(): Unit {
        ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.READ_CALENDAR), 1)
    }

    fun hasPermissionReadCalendar(): Boolean {
        return ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CALENDAR) == PackageManager.PERMISSION_GRANTED
    }
}

